product_state/models/product_template.py

```diff
@@ -1,18 +1,84 @@
 # Copyright 2017 ACSONE SA/NV (<http://acsone.eu>)
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
-from odoo import models, fields
+from odoo import api, models, fields
+
+
+class ProductState(models.Model):
+    _name = 'product.state'
+    _description = "Product State"
+    _order = 'sequence, id'
+
+    name = fields.Char('State Name', required=True, translate=True)
+    code = fields.Char('State Code', required=True)
+    sequence = fields.Integer('Sequence', help="Used to order the States", default=25)
+    description = fields.Text(translate=True)
+    product_ids = fields.One2many(
+        'product.template',
+        'product_state_id',
+        string='State Products',
+    )
+    products_count = fields.Integer(
+        string='Number of products',
+        compute='_compute_products_count',
+    )
+
+    _sql_constraints = [
+        ('product_state_code_unique', 'UNIQUE(code)', 'Product State Code must be unique.')
+        ]
+
+    @api.depends('product_ids')
+    def _compute_products_count(self):
+        data = self.env["product.template"].read_group([
+            ('product_state_id', 'in', self.ids),
+        ], ['product_state_id'], ['product_state_id'])
+        mapped_data = {
+            record['product_state_id'][0]: record['product_state_id_count']
+            for record in data
+        }
+        for state in self:
+            state.products_count = mapped_data.get(state.id, 0)
+
+    def post_init_hook(cr, registry):
+        """ This hook is used to add a state on existing products
+        when module product_state is installed.
+        """
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        product_ids  = env['product.template'].search([])
+        product_with_state = env['product.state'].search([]).mapped('product_state_id')
+        product_without_state = product_ids - product_with_state
+        for product in product_without_state:
+                product_state_id = product_state.id,
+            })
```

Comment:
Where is this function locate in  the stock module?

Reply:
Should be in the __init__.py file?

Reply:
or should I add another hooks.py file?

Reply:
Yes, is it.

Reply:
and before I add commit about function

Reply:
<pre>
product_ids  = env['product.template'].search([])
product_with_state = env['product.state'].search([]).mapped('product_state_id')
product_without_state = product_ids - product_with_state
</pre>
equivalent
<pre>
product_without_state  = env['product.template'].search([('product_state_id', '=', False)])
</pre>

And I don`t uderstand what it is.
<pre>
for product in product_without_state:
                product_state_id = product_state.id,
            })
</pre>
Rewrite 
<pre>
product_without_state._compute_product_state()
</pre>
And check your code before push

Reply:
ok
I can't check the code properly before push. I work in Windows programming in CAD systems (like SolidWorks). 
Actually, I am doing this PR while I am finishing an improvement in our ICE (integration module beteween SolidWorks & Odoo)

![image](https://user-images.githubusercontent.com/24428859/76153854-b9832600-60d2-11ea-94bc-28b7de0a7f62.png)